src/main/java/frc/robot/subsystems/Photonvision.java

```diff
@@ -0,0 +1,168 @@
+import edu.wpi.first.math.geometry.Rotation2d;
+import edu.wpi.first.math.geometry.Transform3d;
+import edu.wpi.first.math.geometry.Translation2d;
+import edu.wpi.first.math.geometry.Translation3d;
+import edu.wpi.first.math.util.Units;
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.Constants.GameConstants;
+import frc.robot.Constants.RobotConstants;
+
+import java.util.ArrayList;
+
+import org.photonvision.PhotonCamera;
+import org.photonvision.PhotonPoseEstimator;
+import org.photonvision.PhotonUtils;
+import org.photonvision.PhotonPoseEstimator.PoseStrategy;
+import org.photonvision.targeting.PhotonPipelineResult;
+import org.photonvision.targeting.PhotonTrackedTarget;
+
+public class Photonvision extends SubsystemBase {
+
+  PhotonCamera camera = new PhotonCamera("OV5647");
+  AprilTagFieldLayout aprilTagFieldLayout;
+
+  ArrayList<Pair<PhotonCamera, Transform3d>> camList = new ArrayList<Pair<PhotonCamera, Transform3d>>();
+  PhotonPoseEstimator poseEstimator = new PhotonPoseEstimator(aprilTagFieldLayout, PoseStrategy.CLOSEST_TO_CAMERA_HEIGHT, camera, RobotConstants.robotToCam);
+
+  public Photonvision() {
+    try {
+      aprilTagFieldLayout = AprilTagFieldLayout.loadFromResource(AprilTagFields.kDefaultField.m_resourceFile);
+    } catch (Exception e) {
+      System.out.println(e);
+    }
+    camList.add(new Pair<PhotonCamera, Transform3d>(camera, RobotConstants.robotToCam));
+    GameConstants.tagMap.put(1, new Double[]{1551.35, 107.16, 46.27, 180.0});
+    GameConstants.tagMap.put(2, new Double[]{1551.35, 274.80, 46.27, 180.0});
+    GameConstants.tagMap.put(3, new Double[]{1551.35, 442.44, 46.27, 180.0});
+    GameConstants.tagMap.put(4, new Double[]{1617.87, 674.97, 69.54, 180.0});
+    GameConstants.tagMap.put(5, new Double[]{36.19, 674.97, 69.54, 0.0});
+    GameConstants.tagMap.put(6, new Double[]{102.743, 442.44, 46.27, 0.0});
+    GameConstants.tagMap.put(7, new Double[]{102.743, 274.80, 46.27, 0.0});
+    GameConstants.tagMap.put(8, new Double[]{102.743, 107.16, 46.27, 0.0});
+
+  }
+
+  public class CameraData {
+    public boolean hasTargets;
+    public double targetPitch;
+    public double targetYaw;
+    public double targetArea;
+    public double targetSkew;
+    public Transform3d targetPose;
+  }
+
+  public class AprilTagData {
+    public int targetId;
+    public double poseAmbiguity;
+    //public Optional<Pose3d> tagPose;
+    public Double[] tagPose2;
+    public Transform3d alternateCameraToTarget;
+  }
+
+  public CameraData data = new CameraData();
+  public AprilTagData tagData = new AprilTagData();
+
+  public int getPipelineIndex() {
+    return camera.getPipelineIndex();
+  }
+
+  public void togglePipeline() {
+    camera.setPipelineIndex( (getPipelineIndex() == 1) ? 2 : 1);
+  }
+
+  public void updateData() {
+    PhotonPipelineResult results = camera.getLatestResult();
+
+    if (results == null) {
```

Comment:
Could be a one-liner, but this is more stylistic, so this can be ignored if you like. Same for lines 94 and 98.